src/error/src/anyhow.rs

```diff
@@ -0,0 +1,106 @@
+// Copyright 2024 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/// Define a newtype wrapper around [`anyhow::Error`].
+///
+/// # Construction
+///
+/// Unlike [`anyhow::Error`], the newtype **CANNOT** be converted from any error
+/// types implicitly. Instead, it can only be converted from [`anyhow::Error`]
+/// by default.
+///
+/// - Users are encouraged to use [`anyhow::Context`] to attach detailed
+///   information to the source error and make it an [`anyhow::Error`] before
+///   converting it to the newtype.
+///
+/// - Or specify the default context for each source error type in this macro
+///   as shown in the example below, which will enable the implicit conversion
+///   `From` the source error type to the newtype. This should not be preferred
+///   in most cases as the context could be less informative, but it is still
+///   useful during refactoring, or if the source error type is believed to be
+///   self-explanatory.
+///
+/// To construct a new error from scratch, one can still use macros like
+/// `anyhow::anyhow!` or `risingwave_common::bail!`. Since `bail!` and `?`
+/// already imply an `into()` call, developers do not need to care about the
+/// type conversion most of the time.
+///
+/// # Example
+///
+/// ```ignore
+/// def_anyhow_newtype! {
+///    /// Documentation for the newtype.
+///    #[derive(..)]
+///    pub MyError,
+///
+///    mysql::Error => "failed to interact with MySQL",
+///    postgres::Error => "failed to interact with PostgreSQL",
+/// }
+/// ```
+///
```

Comment:
It might be better to add call site examples to help understand. (And IMO it's better to put example at the beginning before detailed explanations. (TLDR))

```rust
fn read_mysql() -> Result<(), mysql::Error> {
    ...
}

fn read_redis() -> Result<(), redis::Error> {
    ...
}

fn read() -> Result<(), MyError> {
    ...
    // We cannot directly `?` here without first converting it to anyhow::Error with `.context`
    read_redis().context("failed to read from redis")?;

    // mysql::Error will be automatically converted with `.context("failed to interact with MySQL")`
    read_mysql()?;
    ...
}

```

Reply:
Updated. 🥰